server/src/main/java/org/opensearch/index/translog/RemoteFsTranslog.java

```diff
@@ -313,13 +316,28 @@ protected void setMinSeqNoToKeep(long seqNo) {
         this.minSeqNoToKeep = seqNo;
     }
 
-    @Override
-    void deleteReaderFiles(TranslogReader reader) {
+    void deleteRemoteGeneration(long generation) {
         try {
-            translogTransferManager.deleteTranslog(primaryTermSupplier.getAsLong(), reader.generation);
+            translogTransferManager.deleteTranslog(primaryTermSupplier.getAsLong(), generation);
         } catch (IOException ignored) {
-            logger.error("Exception {} while deleting generation {}", ignored, reader.generation);
+            logger.error("Exception {} while deleting generation {}", ignored, generation);
+        }
+    }
+
+    public void trimUnreferencedReaders() throws IOException {
+        // clean up local translog files and updates readers
+        super.trimUnreferencedReaders();
+
+        // cleans up remote translog files not referenced in latest uploaded metadata
+        // This enables us to restore translog from that metadata
+        for (long generation = minRemoteGenReferenced - 1; generation > 0; generation--) {
+            String translogFilename = Translog.getFilename(generation);
+            if (fileTransferTracker.uploaded(translogFilename)) {
+                logger.trace("delete remote translog generation file [{}], not referenced by metadata anymore", generation);
+                deleteRemoteGeneration(generation);
+            } else {
+                break;
```

Comment:
Why are we breaking out of the for loop?

I can think of that we may want to retry the uploads - in which case should we start the for loop from minimum generation number that exists?

Reply:
this would increase the flush times to a very high number. The clean up can be taken as an async job on its own .